src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,564 @@
+    @assert !hasproperty(newdf, newname)
+    newdf[!, newname] = copycols ? df[:, col_idx] : df[!, col_idx]
+    # mark that column transformation was applied
+    # nothing is not possible otherwise as a value in this dict
+    transformed_cols[newname] = nothing
+end
+
+function select_transform!(nc:: Pair{<:Union{Int, AbstractVector{Int}},
+                                     <:Pair{<:Union{Base.Callable, ByRow}, Symbol}},
+                           df::AbstractDataFrame, newdf::DataFrame,
+                           transformed_cols::Dict{Symbol, Any}, copycols::Bool)
+    col_idx, (fun, newname) = nc
+    @assert !hasproperty(newdf, newname)
+    cdf = eachcol(df)
+    if col_idx isa Int
+        res = fun(df[!, col_idx])
+    else
+        res = fun((cdf[i] for i in col_idx)...)
+    end
+    if res isa Union{AbstractDataFrame, NamedTuple, DataFrameRow, AbstractMatrix}
+        throw(ArgumentError("return value from function $fun " *
+                            "of type $(typeof(res)) is currently not allowed."))
+    end
+    if res isa AbstractVector
+        if copycols && !(fun isa ByRow) && (res isa SubArray ||
+            any(i -> parent(res) === parent(cdf[i]), col_idx))
+            newdf[!, newname] = copy(res)
+        else
+            newdf[!, newname] = res
+        end
+    else
+        newdf[!, newname] = [res]
+    end
+    transformed_cols[newname] = nothing
+end
+
+"""
+    select!(df::DataFrame, inds...)
+
+Mutate `df` in place to retain only columns specified by `inds...` and return it.
+
+Arguments passed as `inds...` can be any index that is allowed for column indexing.
+In particular, regular expressions, `All`, `Between`, and `Not` selectors are supported.
+
+Columns can be renamed using the `old_column => new_column_name` syntax,
+and transformed using the `old_column => fun => new_column_name` syntax.
+`new_column_name` must be a `Symbol`, and `fun` a function or a type. If `old_column`
+is a `Symbol` or an integer then `fun` is applied to the corresponding column vector.
+Otherwise `old_column` can be any column indexing syntax, in which case `fun`
+will be passed the column vectors specified by `old_column` as separate arguments.
+If `fun` returns a value of type other than `AbstractVector` then it will be wrapped
+into a 1-element vector, unless its type is one of `AbstractDataFrame`, `NamedTuple`,
+`DataFrameRow`, `AbstractMatrix`, in which case an error is thrown as currently these
+return types are not allowed.
+
+To apply `fun` to each row instead of whole columns, it can be wrapped in a `ByRow`
+struct. In this case if `old_column` is a `Symbol` or an integer then `fun` is applied
+to each element (row) of `old_column` using broadcasting. Otherwise `old_column` can be
+any column indexing syntax, in which case `fun` will be passed one argument for each of
+the columns specified by `old_column`. If `ByRow` is used it is not allowed that
+`old_column` selects an empty set of columns.
```

Comment:
This sounds like better English (to my non-native speaker ears :-):
```suggestion
the columns specified by `old_column`. If `ByRow` is used it is not allowed for
`old_column` to select an empty set of columns.
```